### PR TITLE
Fix regexp error message in dictionaries

### LINKED ISSUE #14556: Unable to use backslash in the replacement value of dictionaries
### Context

Issue found while while reviewing the list of modified files in SVN update status window.

### Steps to reproduce:

* Open a new entry in the temporary dictionary
* Ine pattern field put "`p:\a\very\long\path\to\my\project\root\folder\`"
* In replace field, put: "`root\`"
* Leave other fields as they are, specifically leave the radio button type on "Anywhere"
* Press OK

### Actual behavior:
I get the following error message box:
> Regular Expression error: "bad escape (end of pattern) at position 4".

### Expected behavior:
* When regular expression is not selected as entry type, I should be able to use backslash character normally in the dictionary replace field.
* I should not get any message related to regexps if I have not selected "Regular expression" as dictionary entry type.

### NVDA logs, crash dumps and other attachments:
Following debugWarning found in the log:
```
DEBUGWARNING - gui.speechDict.DictionaryEntryDialog.onOk (17:27:09.948) - MainThread (19688):
Could not add dictionary entry due to (regex error) : bad escape (end of pattern) at position 4
```

#### NVDA installed/portable/running from source:
Installed
#### NVDA version:
2022.4
#### Windows version:
Windows 10 20H2 (AMD64) build 19042.2364
#### Name and version of other software in use when reproducing the issue:
N/A
#### Other information about your system:
N/A
#### Does the issue still occur after restarting your computer?
Yes
#### Have you tried any other versions of NVDA? If so, please report their behaviors.
Not tested but surely not new.
#### If NVDA add-ons are disabled, is your problem still occurring?
Yes
#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
Not tested;unrelated.

### PR DESCRIPTION
**Pull request by CyrilleB79**
### Link to issue number:
Fixes #14556
### Summary of the issue:

Dictionaries do not allow to use backslash character normally in the replacement field when regexp is not selected as entry type. When such a character is used, a message box complaining of a regexp error appears.

### Description of user facing changes

* The user can use backslash in replacement field normally when entry type is not regexp
* Error messages are more precise and more specific when entering erroneous values in the fields for a regexp entry

### Description of development approach
* When the entry type is not regexp, escape the replacement string
* When entries are created:
  * filter exception on `re.error` instead of the too global `Exception` to catch only exceptions related to regexps
  * separate the try/except statement in two in order to indicate in which field the regexp error lies
  * In case of a regexp error, after validating "OK" in the message box, the field where the error lies is re-focused
  * Reports the regexp error only in case of entry type = regexp; in the other cases,re-raise the exception, in order not to catch unrelated exceptions, e.g. code bug.

### Testing strategy:

Manual tests:

* Add entry with backslash in replacement field and entry type = everywhere
* Add entry with error either in pattern or in replacement field with entry type = regexp.

### Known issues with pull request:
None
### Change log entries:
Bug fixes
`It is now possible to use the backslash character in the replacement field of a dictionaries entry, when the type is not set to regular expression. (#14556)`

### Code Review Checklist:

 [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] Security precautions taken.

### CODE REVIEW

**Comment by XLTechie** [comment]
For changelog:

> It is now again possible to use backslash character in the replacement field of dictionaries. (#14556)

Suggestion:

It is now possible to use the backslash character in the replacement field of dictionaries, when the pattern is not set to regular expression. (#14556)

File: source/gui/speechDict.py [hunk 1]
```diff
@@ -94,20 +96,37 @@ def onOk(self, evt):
...
+				wx.OK | wx.ICON_WARNING,
+				self
 			)
+			return
+		try:
 			dictEntry.sub("test")  # Ensure there are no grouping error (#11407)
-		except Exception as e:
-			log.debugWarning("Could not add dictionary entry due to (regex error) : %s" % e)
+		except RegexpError as e:
+			log.debugWarning("Could not add dictionary entry due to (regex error in the replacement field) : %s" % e)
```
Comment (XLTechie) [inline source/gui/speechDict.py:124]: Why keep the parentheses around the error itself? "(regex error in the replacement field)"

And perhaps I am missing something, but why not switch to F-strings for these messages?

**Comment by CyrilleB79** [comment]
@XLTechie thanks for the review. I have addressed the raised points in the code and in the initial description.

File: source/speechDictHandler/__init__.py [hunk 2]
```diff
@@ -1,8 +1,8 @@
-#speechDictHandler.py
-#A part of NonVisual Desktop Access (NVDA)
-#Copyright (C) 2006-2017 NVDA Contributors <http://www.nvda-project.org/>
-#This file is covered by the GNU General Public License.
-#See the file COPYING for more details.
+# speechDictHandler.py
```
Comment (seanbudd) [inline source/speechDictHandler/__init__.py:1]:
```suggestion
```

File: source/gui/speechDict.py [hunk 3]
```diff
@@ -94,20 +96,37 @@ def onOk(self, evt):
...
-		except Exception as e:
-			log.debugWarning("Could not add dictionary entry due to (regex error) : %s" % e)
+		except RegexpError as e:
+			log.debugWarning(f"Could not add dictionary entry due to regex error in the replacement field : {e}")
+			if entryType != speechDictHandler.ENTRY_TYPE_REGEXP:
+				raise(e)
 			gui.messageBox(
 				# Translators: This is an error message to let the user know that the dictionary entry is not valid.
-				_("Regular Expression error: \"%s\".") % e,
+				_("Regular Expression error in the replacement field: \"{e}\"."),
```
Comment (seanbudd) [inline source/gui/speechDict.py:129]: Missing f in fstring

```suggestion
				_(f"Regular Expression error in the replacement field: \"{e}\"."),
```

**Review by seanbudd — COMMENTED** [review]
Thanks @CyrilleB79

**Comment by CyrilleB79** [comment]
@seanbudd, I have addressed review comments. However, it is not possible to use f-strings with gettext so I have turned back to % interpolation as it was before.

I have also added two more lines to focus the field causing the regexp error as it was already done in case of empty pattern field error.

**Comment by XLTechie** [comment]
> However, it is not possible to use f-strings with gettext so I have turned back to % interpolation as it was before.

I meant only to suggest that for the log messages, not the user messages. 
Apologies if I commented on the wrong line.

**Comment by XLTechie** [comment]
> However, it is not possible to use f-strings with gettext so I have turned back to % interpolation as it was before.

It is possible, however, to use .format() with gettext, if you want to go 
slightly more modern than % replacements in the user messages. See 
globalCommands.py for examples, among others.

**Comment by CyrilleB79** [comment]
> > However, it is not possible to use f-strings with gettext so I have turned back to % interpolation as it was before.
> I meant only to suggest that for the log messages, not the user messages. Apologies if I commented on the wrong line.

No worry, you commented on the correct line. But I have then added another extra f-string in an inappropriate place, i.e. where gettext is used.

Regarding `%` versus `.format()` syntax, is there a recommendation or a preference to use the latter? (NVDA/NVAccess recommendation and/or Python/PEP recommendation)
 I'll let @seanbudd  decide if a change should be done.

**Comment by josephsl** [comment]
Hi, regarding formatted string literals (F strings) and string formatting, the former is not used in translatable strings due to security (translated messages can introduce strings that could crash Python), and using str.format allows translators to see more readable (and understandable) source text for translations. Thanks.

**Comment by seanbudd** [comment]
Agreed with @josephsl, the label in the format string e.g. `{settingsName}` is more explicit and helpful than `%s`

**Comment by CyrilleB79** [comment]
> Agreed with @josephsl, the label in the format string e.g. `{settingsName}` is more explicit and helpful than `%s`

OK @seanbudd, I have fixed it.

File: source/gui/speechDict.py [hunk 4]
```diff
@@ -94,25 +96,44 @@ def onOk(self, evt):
...
 				self.replacementTextCtrl.GetValue(),
 				self.commentTextCtrl.GetValue(),
 				bool(self.caseSensitiveCheckBox.GetValue()),
-				self.getType()
+				entryType,
 			)
+		except RegexpError as e:
+			log.debugWarning(f"Could not add dictionary entry due to regex error in the pattern field : {e}")
+			if entryType != speechDictHandler.ENTRY_TYPE_REGEXP:
+				raise(e)
```
Comment (seanbudd) [inline source/gui/speechDict.py:111]:
```suggestion
				raise e
```

File: source/gui/speechDict.py [hunk 5]
```diff
@@ -94,25 +96,44 @@ def onOk(self, evt):
...
+			self.patternTextCtrl.SetFocus()
+			return
+		try:
 			dictEntry.sub("test")  # Ensure there are no grouping error (#11407)
-		except Exception as e:
-			log.debugWarning("Could not add dictionary entry due to (regex error) : %s" % e)
+		except RegexpError as e:
+			log.debugWarning(f"Could not add dictionary entry due to regex error in the replacement field : {e}")
+			if entryType != speechDictHandler.ENTRY_TYPE_REGEXP:
+				raise(e)
```
Comment (seanbudd) [inline source/gui/speechDict.py:127]:
```suggestion
				raise e
```

**Review by seanbudd — COMMENTED** [review]
Looks good to me, other than a minor syntax issue

**Review by seanbudd — APPROVED** [review]

File: source/speechDictHandler/__init__.py [hunk 6]
```diff
@@ -42,7 +41,10 @@ def __init__(self, pattern, replacement,comment,caseSensitive=True,type=ENTRY_TY
 		self.type=type
 
 	def sub(self, text):
```
Comment (seanbudd) [inline source/speechDictHandler/__init__.py:43]:
```suggestion
	def sub(self, text: str) -> str:
```

**Pull request merged**